src/Analyser/MutatingScope.php

```diff
@@ -2680,6 +2637,86 @@ private function resolveType(Expr $node): Type
 		return new MixedType();
 	}
 
+	private function resolveConcatType(Expr\BinaryOp\Concat|Expr\AssignOp\Concat $node): Type
+	{
+		if ($node instanceof Node\Expr\AssignOp) {
+			$left = $node->var;
+			$right = $node->expr;
+		} else {
+			$left = $node->left;
+			$right = $node->right;
+		}
+
+		$leftStringType = $this->getType($left)->toString();
+		$rightStringType = $this->getType($right)->toString();
+		if (TypeCombinator::union(
+			$leftStringType,
+			$rightStringType,
+		) instanceof ErrorType) {
+			return new ErrorType();
+		}
+
+		if ($leftStringType instanceof ConstantStringType && $leftStringType->getValue() === '') {
+			return $rightStringType;
+		}
+
+		if ($rightStringType instanceof ConstantStringType && $rightStringType->getValue() === '') {
+			return $leftStringType;
+		}
+
+		if ($leftStringType instanceof ConstantStringType && $rightStringType instanceof ConstantStringType) {
+			return $leftStringType->append($rightStringType);
+		}
+
+		// we limit the number of union-types for performance reasons
+		if ($leftStringType instanceof UnionType && count($leftStringType->getTypes()) <= 16 && $rightStringType instanceof ConstantStringType) {
+			$constantStrings = TypeUtils::getConstantStrings($leftStringType);
+			if (count($constantStrings) > 0) {
+				$strings = [];
+				foreach ($constantStrings as $constantString) {
+					if ($constantString->getValue() === '') {
+						$strings[] = $rightStringType;
+
+						continue;
+					}
+					$strings[] = $constantString->append($rightStringType);
+				}
+				return TypeCombinator::union(...$strings);
+			}
+		}
+		if ($rightStringType instanceof UnionType && count($rightStringType->getTypes()) <= 16 && $leftStringType instanceof ConstantStringType) {
+			$constantStrings = TypeUtils::getConstantStrings($rightStringType);
+			if (count($constantStrings) > 0) {
+				$strings = [];
+				foreach ($constantStrings as $constantString) {
+					if ($constantString->getValue() === '') {
+						$strings[] = $leftStringType;
+
+						continue;
+					}
+					$strings[] = $leftStringType->append($constantString);
+				}
+				return TypeCombinator::union(...$strings);
+			}
+		}
```

Comment:
these are the new lines. everything else was moved 1:1 from `resolveType` into this new method.
I figured it would make things a bit better to read in the future.. hopefully you are fine with it.